keps/sig-api-machinery/00xx-publish-crd-openapi.md

```diff
@@ -0,0 +1,142 @@
+
+### Non-Goals
+
+* Support/honor OpenAPI AdditionalProperties in kube-openapi and kubectl, to
+  enable client-side validation on native Kubernetes ObjectMeta and TypeMeta
+  while allowing arbitrary extra fields in a request.
+* Finalize the design on how we aggregate OpenAPI documentation from
+  apiextensions-apiserver
+  - [Short term design doc](https://docs.google.com/document/d/13lBj8Stdwku8BgL0fbT__4Iw97NRh77loJ_MoZuCGwQ/edit#)
+    including multiple workarounds has been purposed and is under discussion.
+  - Long term design doc including generalizing APIService API or implementing
+    watch mechanism for /openapi/v2 endpoint on apiextension-apiserver is work
+    in progress.
+
+## Proposal
+
+ref: https://github.com/kubernetes/kubernetes/pull/71192
+
+### Risks and Mitigations
+
+TODO(roycaihw):
+
+if somehow the OpenAPI documentation for native APIs gets removed, or the code aggregates a huge
+number/size of OpenAPI documentation because of a bug? kube-aggregator might
+(not likely with sttts' optimization) get overwhelmed? kubectl (and anything else that downloads
+openapi from a cluster) might (likely) get overwhelmed?
+
+are we under potential risk of leaking information from the CRD object?
+
+TODO(roycaihw): also do we need to test the behavior above?
+
+mitigate:
+
+currently we use the same feature flag as we use for the json-schema:
+CustomResourceValidation. the feature flag can be turned off to stop
+openapi creation & aggregation & publishing from happening. should we use
+another feature flag separately?
+
+## Graduation Criteria
+
+In 1.14, publishing CRD OpenAPI should be e2e-tested to ensure:
+
+* for CRD with validation schema
+
+1. `kubectl create` client-side validation works to:
+  - allow requests with known and required properites
+  - reject requests with unknown properties
+  - reject requests without required properties
+
+2. `kubectl explain` works to:
+  - explain CRD properties
+  - explain CRD properties recursively
+  - return error when explain is called on property that doesn't exist
+
+* for CRD without validation schema
+
+1. `kubectl create` client-side validation works to:
+  - allow requests with any unknown properties
+
+2. `kubectl explain` works to:
+  - return error NOT FOUND when explain is called for given CRD
+  - TODO(roycaihw): it is also suggested that we can express the definition as
+    an empty object to allow arbitrary properties, and `kubectl explain` won't
+    return error
+
+* for multiple CRDs
+  - CRDs in different groups (two CRDs) show up in OpenAPI documentation
+  - CRDs in the same group but different version (one multiversion CRD or two
+    CRDs) show up in OpenAPI
+    documentation
+  - CRDs in the same group version but different kind (two CRDs) show up in OpenAPI
+    documentation
+
+* for publishing speed
+  - TODO(roycaihw): OpenAPI Paths and Definitions get updated within 10s (magic number) after
+    CRD gets created/updated/deleted?
```

Comment:
it's not only the delay, it is the cpu load which becomes critical. Is 1sec one core cpu load ok?

Reply:
suggest adding "does not measurably increase server load at steady state"

Reply:
added

currently downloading spec and etag from apiextensions-apiserver is in-process function call. We skip aggregation when etag doesn't change. So we are not adding much load